Typescript Conversion: main classes

brisberg opened the pull request:
This is a reasonable start, as the tests still pass and we get some type safety on some things.

Unfortunately I have punted some of the work and left in a large number of `any` types in here to get them to compile. Especially around `@screeps/driver`, `@screeps/common`, `@screeps/storage`, and any of the results from a `db` lookup.

I don't think this project is the right place to put bespoke typings for those main projects, but it will take some time to successfully type them. Maybe mongodb typings would help on the database?

--

As discussed in #7, it might make more sense to target this PR into a `typescript` branch, so others can contribute to the typings.

Also we need to convert the tests to TS as well, but that should be straight forward.

All of the options in the `tsconfig.json` are up in the air. I don't feel strongly about any of them, though this configuration does compile and test /shrug.

pyrodogg commented on src/user.ts:
```diff
@@ -1,13 +1,18 @@
         Constructor
     */
-    constructor(server, data) {
+    constructor(server: any, data: any) {
```
```server: ScreepsServer```

pyrodogg commented on src/world.ts:
```diff
@@ -1,16 +1,28 @@
         Constructor
     */
-    constructor(server) {
+    constructor(server: any) {
```
```server: ScreepsServer```

pyrodogg commented on tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+    "lib": [
+      "esnext"
+    ],
+    "target": "ES6",
```
Change target to be at least ES2017, or higher. With ES6, it's polyfilling Promises which isn't necessary.
```suggestion
    "target": "ES2017",
```

brisberg replied:
Seems to compile/test with ESNext, does that sound good?

brisberg replied:
Seems to compile/test with ESNext. Does that sound good?

brisberg commented:
Feel free to merge if `ESNext` is an acceptable target.

brisberg commented:
Actually, one last thing to review. I removed the `source` flag and added the `declaration` flag to the tsconfig options. This way the compiler will emit a `.d.ts` typings file for the library. Would you prefer to keep the source maps? I don't think they will be needed by consumers on NPM.
